dht_rt_seeder.go

```diff
@@ -0,0 +1,119 @@
+package dht
+
+import (
+	"context"
+	"time"
+
+	"github.com/ipfs/go-todocounter"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// SeedDialGracePeriod is the grace period for one dial attempt
+var SeedDialGracePeriod = 5 * time.Second
+
+// SeederDialTimeout is the total grace period for a group of dial attempts
+var SeederDialTimeout = 30 * time.Second
+
+// SeederConcurrentDials is the number of peers we will dial simultaneously
+var SeederConcurrentDials = 50
+
+func (dht *IpfsDHT) seedRoutingTable(candidates, fallbacks []peer.ID) error {
+	peersFromSet := func(m map[peer.ID]struct{}) []peer.ID {
+		var peers []peer.ID
+		for p := range m {
+			peers = append(peers, p)
+		}
+
+		return peers
+	}
+
+	// candidates we have still not tried dialling to
+	remainingCandidates := make(map[peer.ID]struct{})
+	// fallbacks we have still not tried dialling to
+	remainingFallbacks := make(map[peer.ID]struct{})
+
+	for i := range candidates {
+		remainingCandidates[candidates[i]] = struct{}{}
+	}
+
+	for i := range fallbacks {
+		remainingFallbacks[fallbacks[i]] = struct{}{}
+	}
+
+	for {
+
+		todial := dht.seedsProposer.Propose(dht.routingTable, peersFromSet(remainingCandidates), peersFromSet(remainingFallbacks))
+		if len(todial) == 0 {
+			logger.Infof("dht rt seeder: finished as proposer returned nil, RT size is now %d", dht.routingTable.Size())
+			return nil
+		} else {
+			// start dialing
+			type result struct {
+				p   peer.ID
+				err error
+			}
+
+			// attempts to dial to a given peer to verify it's available
+			dialFn := func(ctx context.Context, p peer.ID, res chan<- result) {
+				childCtx, cancel := context.WithTimeout(ctx, SeedDialGracePeriod)
+				defer cancel()
+				_, err := dht.host.Network().DialPeer(childCtx, p)
+				select {
+				case <-ctx.Done(): // caller has already hung up & gone away
+				case res <- result{p, err}:
+				}
+			}
+
+			resCh := make(chan result) // dial results.
+			ctx, cancel := context.WithTimeout(dht.ctx, SeederDialTimeout)
+			defer cancel()
+
+			// start dialing
+			semaphore := make(chan struct{}, SeederConcurrentDials)
+			go func(peers []peer.ID) {
+				for _, p := range peers {
+					semaphore <- struct{}{}
+					go func(p peer.ID, res chan<- result) {
+						dialFn(ctx, p, resCh)
+						<-semaphore
+					}(p, resCh)
+				}
+			}(todial)
+
+			// number of results we expect on the result channel
+			todo := todocounter.NewSyncCounter()
```

Comment:
Why not use sync.WaitGroup or just an int. This isn't threaded afaict.

Reply:
Yes, we could use a simple int here to keep a track of remaining Dials & exit the loop when it goes to zero. A `todocounter` really shines when the go-routines we create for tasks(dial attempts for example) create more tasks themselves.

https://github.com/ipfs/go-todocounter/blob/742667602a47ab3a2b7f17d935019c3255719dce/counter.go#L7

But, we do NOT have that here as a dial attempt does not generate one more peer we need to dial to.